File: ios/Sources/PhotoLibraryPlugin/PhotoLibraryPlugin.swift
import Capacitor
import Foundation
import Photos
import PhotosUI

@objc(PhotoLibraryPlugin)
public class PhotoLibraryPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion: String = "8.0.8"
    public let identifier = "PhotoLibraryPlugin"
    public let jsName = "PhotoLibrary"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "checkAuthorization", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestAuthorization", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAlbums", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getLibrary", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPhotoUrl", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getThumbnailUrl", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pickMedia", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private var service = PhotoLibraryService()
    private var pendingPickCall: CAPPluginCall?
    private var pendingPickOptions: PhotoLibraryPickOptions?

    override public func load() {
        super.load()
        service.webPathResolver = { [weak self] localURL in
            guard let portable = self?.bridge?.portablePath(fromLocalURL: localURL) else {
                return nil
            }
            return portable.absoluteString
        }
        service.prepareCacheDirectories()
    }

    @objc public func checkAuthorization(_ call: CAPPluginCall) {
        let state = service.currentAuthorizationState()
        call.resolve(["state": state.rawValue])
    }

    @objc public func requestAuthorization(_ call: CAPPluginCall) {
        service.requestAuthorization { state in
            call.resolve(["state": state.rawValue])
        }
    }

    @objc public func getAlbums(_ call: CAPPluginCall) {
        guard service.isAccessGranted else {
            call.reject(PhotoLibraryError.permissionDenied.message)
            return
        }

        service.fetchAlbums { [weak self] albums in
            guard let self = self else { return }
            let data = albums.map { $0.toDictionary(resolver: self.service.webPathResolver) }
            call.resolve(["albums": data])
        }
    }

    @objc public func getLibrary(_ call: CAPPluginCall) {
        guard service.isAccessGranted else {
            call.reject(PhotoLibraryError.permissionDenied.message)
            return
        }

        do {
            let options = try PhotoLibraryGetLibraryOptions(from: call)
            service.fetchLibrary(options: options) { [weak self] result in
                guard let self = self else { return }
                let assets = result.assets.map { $0.toDictionary(resolver: self.service.webPathResolver) }
                call.resolve([
                    "assets": assets,
                    "totalCount": result.totalCount,
                    "hasMore": result.hasMore
                ])
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc public func getPhotoUrl(_ call: CAPPluginCall) {
        guard service.isAccessGranted else {
            call.reject(PhotoLibraryError.permissionDenied.message)
            return
        }

        guard let id = call.getString("id"), !id.isEmpty else {
            call.reject("Parameter 'id' is required")
            return
        }

        service.fetchFullResolutionFile(for: id, allowNetworkAccess: true) { [weak self] file in
            guard let self = self else { return }
            guard let file = file else {
                call.reject(PhotoLibraryError.assetNotFound.message)
                return
            }
            call.resolve(file.toDictionary(resolver: self.service.webPathResolver))
        }
    }

    @objc public func getThumbnailUrl(_ call: CAPPluginCall) {
        guard service.isAccessGranted else {
            call.reject(PhotoLibraryError.permissionDenied.message)
            return
        }

        guard let id = call.getString("id"), !id.isEmpty else {
            call.reject("Parameter 'id' is required")
            return
        }

        let width = call.getInt("width") ?? PhotoLibraryDefaults.thumbnailWidth
        let height = call.getInt("height") ?? PhotoLibraryDefaults.thumbnailHeight
        let quality = call.getDouble("quality") ?? PhotoLibraryDefaults.thumbnailQuality

        service.fetchThumbnailFile(for: id, width: width, height: height, quality: quality, allowNetworkAccess: true) { [weak self] file in
            guard let self = self else { return }
            guard let file = file else {
                call.reject(PhotoLibraryError.assetNotFound.message)
                return
            }
            call.resolve(file.toDictionary(resolver: self.service.webPathResolver))
        }
    }

    @objc public func pickMedia(_ call: CAPPluginCall) {
        guard #available(iOS 14, *) else {
            call.reject("pickMedia requires iOS 14 or newer")
            return
        }

        DispatchQueue.main.async {
            if self.pendingPickCall != nil {
                call.reject("Another pickMedia call is already in progress")
                return
            }

            do {
                let options = try PhotoLibraryPickOptions(from: call)
                var configuration = PHPickerConfiguration(photoLibrary: PHPhotoLibrary.shared())
                configuration.selectionLimit = options.selectionLimit
                var filters: [PHPickerFilter] = []
                if options.includeImages {
                    filters.append(.images)
                }
                if options.includeVideos {
                    filters.append(.videos)
                }
                if filters.isEmpty {
                    filters = [.images]
                }
                configuration.filter = PHPickerFilter.any(of: filters)

                let picker = PHPickerViewController(configuration: configuration)
                picker.delegate = self

                guard let presenter = self.bridge?.viewController else {
                    call.reject("Unable to access view controller to present picker")
                    return
                }

                self.pendingPickCall = call
                self.pendingPickOptions = options
                presenter.present(picker, animated: true)
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }
}

@available(iOS 14, *)
extension PhotoLibraryPlugin: PHPickerViewControllerDelegate {
    public func picker(_ picker: PHPickerViewController, didFinishPicking results: [PHPickerResult]) {
        picker.dismiss(animated: true)

        guard let call = pendingPickCall else {
            return
        }

        let options = pendingPickOptions
        pendingPickCall = nil
        pendingPickOptions = nil

        guard let options = options else {
            call.resolve(["assets": []])
            return
        }

        service.processPickerResults(results, options: options) { [weak self] assets in
            guard let self = self else { return }
            let mapped = assets.map { $0.toDictionary(resolver: self.service.webPathResolver) }
            call.resolve(["assets": mapped])
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": self.pluginVersion])
    }

}

File: android/src/main/java/app/capgo/plugin/photo_library/PhotoLibraryPlugin.java
package app.capgo.plugin.photo_library;

import android.Manifest;
import android.os.Build;
import androidx.annotation.NonNull;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@CapacitorPlugin(
    name = "PhotoLibrary",
    permissions = {
        @Permission(
            strings = { Manifest.permission.READ_MEDIA_IMAGES, Manifest.permission.READ_MEDIA_VIDEO },
            alias = PhotoLibraryPlugin.PERMISSION_MEDIA
        ),
        @Permission(strings = { Manifest.permission.READ_EXTERNAL_STORAGE }, alias = PhotoLibraryPlugin.PERMISSION_MEDIA_LEGACY)
    }
)
public class PhotoLibraryPlugin extends Plugin {

    static final String PERMISSION_MEDIA = "media";
    static final String PERMISSION_MEDIA_LEGACY = "media_legacy";

    private static final String STATE_AUTHORIZED = "authorized";
    private static final String STATE_DENIED = "denied";

    private final ExecutorService executor = Executors.newFixedThreadPool(2);
    private PhotoLibraryService service;

    @Override
    public void load() {
        super.load();
        service = new PhotoLibraryService(getContext(), getBridge());
        service.prepareCacheDirectories();
    }

    @Override
    protected void handleOnDestroy() {
        super.handleOnDestroy();
        executor.shutdown();
        service = null;
    }

    @PluginMethod
    public void checkAuthorization(PluginCall call) {
        call.resolve(statusObject(currentAuthorizationState()));
    }

    @PluginMethod
    public void requestAuthorization(PluginCall call) {
        String alias = permissionAlias();
        if (hasPermission(alias)) {
            call.resolve(statusObject(STATE_AUTHORIZED));
            return;
        }

        bridge.saveCall(call);
        requestPermissionForAlias(alias, call, "permissionCallback");
    }

    @PermissionCallback
    private void permissionCallback(PluginCall call) {
        String state = hasMediaPermissions() ? STATE_AUTHORIZED : STATE_DENIED;
        call.resolve(statusObject(state));
    }

    @PluginMethod
    public void getAlbums(PluginCall call) {
        if (!hasMediaPermissions()) {
            call.reject(PhotoLibraryService.PERMISSION_ERROR);
            return;
        }

        executor.execute(() -> {
            try {
                JSArray albums = service.fetchAlbums();
                JSObject result = new JSObject();
                result.put("albums", albums);
                call.resolve(result);
            } catch (Exception ex) {
                call.reject(ex.getMessage(), ex);
            }
        });
    }

    @PluginMethod
    public void getLibrary(PluginCall call) {
        if (!hasMediaPermissions()) {
            call.reject(PhotoLibraryService.PERMISSION_ERROR);
            return;
        }

        GetLibraryOptions options;
        try {
            options = GetLibraryOptions.fromCall(call);
        } catch (IllegalArgumentException ex) {
            call.reject(ex.getMessage());
            return;
        }

        executor.execute(() -> {
            try {
                PhotoLibraryFetchResult result = service.fetchLibrary(options);
                JSObject payload = new JSObject();
                payload.put("assets", result.assets);
                payload.put("totalCount", result.totalCount);
                payload.put("hasMore", result.hasMore);
                call.resolve(payload);
            } catch (Exception ex) {
                call.reject(ex.getMessage(), ex);
            }
        });
    }

    @PluginMethod
    public void getPhotoUrl(PluginCall call) {
        if (!hasMediaPermissions()) {
            call.reject(PhotoLibraryService.PERMISSION_ERROR);
            return;
        }

        String id = call.getString("id");
        if (id == null || id.isEmpty()) {
            call.reject("Parameter 'id' is required");
            return;
        }

        executor.execute(() -> {
            try {
                JSObject file = service.getFullResolutionFile(id);
                if (file == null) {
                    call.reject(PhotoLibraryService.ASSET_NOT_FOUND);
                    return;
                }
                call.resolve(file);
            } catch (Exception ex) {
                call.reject(ex.getMessage(), ex);
            }
        });
    }

    @PluginMethod
    public void getThumbnailUrl(PluginCall call) {
        if (!hasMediaPermissions()) {
            call.reject(PhotoLibraryService.PERMISSION_ERROR);
            return;
        }

        String id = call.getString("id");
        if (id == null || id.isEmpty()) {
            call.reject("Parameter 'id' is required");
            return;
        }

        int width = call.getInt("width", PhotoLibraryDefaults.THUMBNAIL_WIDTH);
        int height = call.getInt("height", PhotoLibraryDefaults.THUMBNAIL_HEIGHT);
        double quality = call.getDouble("quality", PhotoLibraryDefaults.THUMBNAIL_QUALITY);

        executor.execute(() -> {
            try {
                JSObject file = service.getThumbnailFile(id, width, height, quality);
                if (file == null) {
                    call.reject(PhotoLibraryService.ASSET_NOT_FOUND);
                    return;
                }
                call.resolve(file);
            } catch (Exception ex) {
                call.reject(ex.getMessage(), ex);
            }
        });
    }

    private JSObject statusObject(@NonNull String state) {
        JSObject result = new JSObject();
        result.put("state", state);
        return result;
    }

    private boolean hasMediaPermissions() {
        return hasPermission(permissionAlias());
    }

    private String currentAuthorizationState() {
        return hasMediaPermissions() ? STATE_AUTHORIZED : STATE_DENIED;
    }

    private String permissionAlias() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU ? PERMISSION_MEDIA : PERMISSION_MEDIA_LEGACY;
    }
}
